examples/src/main/scala/org/apache/spark/examples/ml/ComplexPipelineExample.scala

```diff
@@ -0,0 +1,156 @@
+import org.apache.spark.ml.feature.{HashingTF, StringIndexer, Tokenizer}
+import org.apache.spark.mllib.evaluation.MulticlassMetrics
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.hive.HiveContext
+import org.apache.spark.{SparkConf, SparkContext}
+
+/**
+ * An example of an end to end machine learning pipeline that classifies text
+ * into one of twenty possible news categories. The dataset is the 20newsgroups
+ * dataset (http://qwone.com/~jason/20Newsgroups/20news-bydate.tar.gz)
+ *
+ * We assume some minimal preprocessing of this dataset has been done to unzip the dataset and
+ * load the data into HDFS as follows:
+ *  wget http://qwone.com/~jason/20Newsgroups/20news-bydate.tar.gz
+ *  tar -xvzf 20news-bydate.tar.gz
+ *  hadoop fs -mkdir ${20news.root.dir}
+ *  hadoop fs -copyFromLocal 20news-bydate-train/ ${20news.root.dir}
+ *  hadoop fs -copyFromLocal 20news-bydate-test/ ${20news.root.dir}
+ *
+ * This example uses Hive to schematize the data as tables, in order to map the folder
+ * structure ${20news.root.dir}/{20news-bydate-train, 20news-bydate-train}/{newsgroup}/
+ * to partition columns type, newsgroup resulting in a dataset with three columns:
+ *  type, newsgroup, text
+ *
+ * In order to run this example, Spark needs to be build with hive, and at runtime there
+ * should be a valid hive-site.xml in $SPARK_HOME/conf with at minimal the following
+ * configuration:
+ * <configuration>
+ *   <property>
+ *     <name>hive.metastore.uris</name>
+ * <!-- Ensure that the following statement points to the Hive Metastore URI in your cluster -->
+ *     <value>thrift://${thriftserver.host}:${thriftserver.port}</value>
+ *   <description>URI for client to contact metastore server</description>
+ *   </property>
+ * </configuration>
+ *
+ * Run with
+ * {{{
+ * bin/spark-submit --class org.apache.spark.examples.ml.ComplexPipelineExample
+ *   --driver-memory 4g [examples JAR path] ${20news.root.dir}
+ * }}}
+ */
+object ComplexPipelineExample {
+
+  def main(args: Array[String]): Unit = {
+    val conf = new SparkConf().setAppName("ComplexPipelineExample")
+    val sc = new SparkContext(conf)
+    val sqlContext = new HiveContext(sc)
+    val path = args(0)
+
+    sqlContext.sql(s"""CREATE EXTERNAL TABLE IF NOT EXISTS 20NEWS(text String)
+      PARTITIONED BY (type String, newsgroup String)
+      STORED AS TEXTFILE location '$path'""")
+
+    val newsgroups = Array("alt.atheism", "comp.graphics",
+      "comp.os.ms-windows.misc", "comp.sys.ibm.pc.hardware",
+      "comp.sys.mac.hardware", "comp.windows.x", "misc.forsale",
+      "rec.autos", "rec.motorcycles", "rec.sport.baseball",
+      "rec.sport.hockey", "sci.crypt", "sci.electronics",
+      "sci.med", "sci.space", "soc.religion.christian",
+      "talk.politics.guns", "talk.politics.mideast",
+      "talk.politics.misc", "talk.religion.misc")
+
+    for (t <- Array("20news-bydate-train", "20news-bydate-train")) {
```

Comment:
I assume one of these should be "test"

A comment on what this is doing would help